File: 1547-minimum-cost-to-cut-a-stick/1547-minimum-cost-to-cut-a-stick.cpp
class Solution {
public:
    int dp[102][102]={};
    int minCost(int n, vector<int>& cuts) {
        cuts.push_back(0);
        cuts.push_back(n);
        sort(cuts.begin(),cuts.end());
        int c=cuts.size();
        for(int i=c-2;i>=1;i--)
        {
            for(int j=1;j<=c-2;j++)
            {
                if(i>j)
                    continue;
                    int minCost=INT_MAX;
                    for(int index=i;index<=j;index++)
                    {
                        int cost=cuts[j+1]-cuts[i-1]+dp[i][index-1]+dp[index+1][j];
                        minCost=min(cost,minCost);
                    }
                dp[i][j]=minCost;
            }
        }
        return dp[1][c-2];
    }
};

File: 1547-minimum-cost-to-cut-a-stick/1547-minimum-cost-to-cut-a-stick.py
class Solution:
    def minCost(self, n: int, cuts: List[int]) -> int:
        cuts.extend([0,n])
        cuts.sort()
        dp = [[0]*len(cuts) for _ in cuts]
        
        for i in range(len(cuts) - 1, -1, -1):
            for j in range(i+2, len(cuts)):
                dp[i][j] = cuts[j] - cuts[i] + min(dp[i][k] + dp[k][j] for k in range(i+1, j))
        
        return dp[0][-1]